File: src/imagery/i.hyper/libsixsv/examples/03_retrievals.c
/**
 * 03_retrievals.c — Atmospheric state retrievals.
 *
 * Demonstrates:
 *  - Solar position (sixs_possol)
 *  - Surface pressure from elevation (ISA)
 *  - H₂O column from 940 nm band depth (Kaufman & Gao 1992)
 *  - AOD from MODIS DDV + MAIAC patch-median smoothing
 *  - O₃ from Chappuis 600 nm band depth
 *  - Cloud/shadow/water/snow quality bitmask
 *  - Joint AOD + H₂O optimal estimation (ISOFIT-inspired)
 *
 * Compile:
 *   gcc -std=c11 -O2 -fopenmp -I../include 03_retrievals.c \
 *       -L../testsuite -lsixsv -lm -fopenmp -Wl,-rpath,../testsuite -o 03_retrievals
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>
#include <string.h>
#include "atcorr.h"
#include "retrieve.h"
#include "oe_invert.h"

/* Synthetic band data helpers */
static void fill_band(float *b, int n, float base, float noise) {
    for (int i = 0; i < n; i++)
        b[i] = base + noise * sinf(i * 0.37f);
}

int main(void)
{
    const int nrows = 32, ncols = 32, npix = nrows * ncols;
    const int doy = 185;   /* summer aphelion */
    const float sza = 32.0f, vza = 5.0f;

    /* ── 1. Solar position ───────────────────────────────────────────────── */
    float sol_zen, sol_az;
    sixs_possol(/*month=*/7, /*day=*/4, /*UTC=*/10.5f,
                /*lon=*/-122.0f, /*lat=*/37.5f,
                &sol_zen, &sol_az, /*year=*/2024);
    printf("Solar position: zenith = %.2f°  azimuth = %.2f°\n", sol_zen, sol_az);

    /* ── 2. Surface pressure from elevation ──────────────────────────────── */
    float p_surface = retrieve_pressure_isa(1500.0f);  /* 1500 m elevation */
    printf("Surface pressure at 1500 m: %.1f hPa\n", p_surface);

    /* ── 3. Synthetic per-pixel radiances ────────────────────────────────── */
    float *L_470  = malloc(npix * sizeof(float));
    float *L_540  = malloc(npix * sizeof(float));
    float *L_600  = malloc(npix * sizeof(float));
    float *L_660  = malloc(npix * sizeof(float));
    float *L_680  = malloc(npix * sizeof(float));
    float *L_860  = malloc(npix * sizeof(float));
    float *L_865  = malloc(npix * sizeof(float));
    float *L_940  = malloc(npix * sizeof(float));
    float *L_1040 = malloc(npix * sizeof(float));
    float *L_2130 = malloc(npix * sizeof(float));

    fill_band(L_470,  npix, 80.0f,  5.0f);
    fill_band(L_540,  npix, 70.0f,  3.0f);
    fill_band(L_600,  npix, 62.0f,  3.0f);
    fill_band(L_660,  npix, 55.0f,  4.0f);
    fill_band(L_680,  npix, 50.0f,  3.0f);
    fill_band(L_860,  npix, 120.0f, 8.0f);
    fill_band(L_865,  npix, 118.0f, 8.0f);
    fill_band(L_2130, npix, 25.0f,  2.0f);
    /* 940 nm: attenuated by water vapour (simulated ~15% depth) */
    for (int i = 0; i < npix; i++)
        L_940[i]  = L_865[i] * 0.85f - 2.0f * sinf(i * 0.5f);
    for (int i = 0; i < npix; i++)
        L_1040[i] = L_865[i] * 0.92f;

    /* ── 4. H₂O retrieval from 940 nm band depth ─────────────────────────── */
    float *wvc = malloc(npix * sizeof(float));
    retrieve_h2o_940(L_865, L_940, L_1040,
                     /*fwhm_um=*/0.0f,   /* broadband — use default K */
                     npix, sza, vza, wvc);
    float wvc_mean = 0.0f;
    for (int i = 0; i < npix; i++) wvc_mean += wvc[i];
    printf("\nH₂O retrieval (940 nm): mean WVC = %.2f g/cm²\n", wvc_mean / npix);

    /* ── 5. AOD retrieval (DDV) ───────────────────────────────────────────── */
    float *aod_px = malloc(npix * sizeof(float));
    float aod_scene = retrieve_aod_ddv(L_470, L_660, L_860, L_2130,
                                       npix, doy, sza, aod_px);
    printf("AOD retrieval (DDV): scene mean = %.3f\n", aod_scene);

    /* MAIAC patch-median spatial regularisation (32×32 patch) */
    retrieve_aod_maiac(aod_px, nrows, ncols, /*patch_sz=*/8);
    float aod_mean = 0.0f;
    for (int i = 0; i < npix; i++) aod_mean += aod_px[i];
    printf("AOD after MAIAC smoothing: mean = %.3f\n", aod_mean / npix);

    /* ── 6. O₃ retrieval from Chappuis band ──────────────────────────────── */
    float o3_du = retrieve_o3_chappuis(L_540, L_600, L_680, npix, sza, vza);
    printf("O₃ column (Chappuis 600 nm): %.0f DU\n", o3_du);

    /* ── 7. Quality bitmask ───────────────────────────────────────────────── */
    uint8_t *mask = malloc(npix * sizeof(uint8_t));
    retrieve_quality_mask(L_470, L_660, L_860,
                          /*L_swir=*/NULL,     /* skip snow detection */
                          npix, doy, sza, mask);
    int n_cloud = 0, n_water = 0;
    for (int i = 0; i < npix; i++) {
        if (mask[i] & RETRIEVE_MASK_CLOUD) n_cloud++;
        if (mask[i] & RETRIEVE_MASK_WATER) n_water++;
    }
    printf("Quality mask: cloud=%d  water=%d  (of %d pixels)\n",
           n_cloud, n_water, npix);

    /* ── 8. Joint AOD + H₂O optimal estimation ────────────────────────────── */
    float wl_vis[] = {0.47f, 0.55f, 0.66f, 0.87f};
    int   n_vis    = 4;
    float aod_lut[] = {0.0f, 0.1f, 0.2f, 0.4f};
    float h2o_lut[] = {1.0f, 2.0f, 3.0f};
    int   N = 4 * 3 * n_vis;

    float *Ra = calloc(N, sizeof(float));
    float *Td = calloc(N, sizeof(float));
    float *Tu = calloc(N, sizeof(float));
    float *sa = calloc(N, sizeof(float));
    LutConfig cfg_oe = {
        .wl = wl_vis, .n_wl = n_vis,
        .aod = aod_lut, .n_aod = 4,
        .h2o = h2o_lut, .n_h2o = 3,
        .sza = sza, .vza = vza, .raa = 90.0f,
        .altitude_km   = 1000.0f, .atmo_model = ATMO_US62,
        .aerosol_model = AEROSOL_CONTINENTAL, .ozone_du = 300.0f,
    };
    LutArrays lut_oe = { .R_atm=Ra, .T_down=Td, .T_up=Tu, .s_alb=sa };
    atcorr_compute_lut(&cfg_oe, &lut_oe);

    /* Simulated TOA reflectances [npix × n_vis] interleaved */
    float *rho_vis = malloc((size_t)npix * n_vis * sizeof(float));
    for (int i = 0; i < npix; i++)
        for (int b = 0; b < n_vis; b++)
            rho_vis[i * n_vis + b] = 0.10f + 0.05f * b + 0.02f * sinf(i * 0.2f);

    float *out_aod = malloc(npix * sizeof(float));
    float *out_h2o = malloc(npix * sizeof(float));
    oe_invert_aod_h2o(&cfg_oe, &lut_oe,
                      rho_vis, npix, n_vis, wl_vis,
                      L_865, L_940, L_1040,
                      sza, vza,
                      /*aod_prior=*/aod_scene,
                      /*h2o_prior=*/wvc_mean / npix,
                      /*sigma_aod=*/0.5f, /*sigma_h2o=*/1.0f,
                      /*sigma_spec=*/0.01f, /*fwhm_940_um=*/0.0f,
                      out_aod, out_h2o);

    float mean_aod = 0.0f, mean_h2o = 0.0f;
    for (int i = 0; i < npix; i++) {
        mean_aod += out_aod[i];
        mean_h2o += out_h2o[i];
    }
    printf("OE retrieval: mean AOD = %.3f  mean H₂O = %.2f g/cm²\n",
           mean_aod / npix, mean_h2o / npix);

    free(L_470); free(L_540); free(L_600); free(L_660); free(L_680);
    free(L_860); free(L_865); free(L_940); free(L_1040); free(L_2130);
    free(wvc); free(aod_px); free(mask);
    free(Ra); free(Td); free(Tu); free(sa);
    free(rho_vis); free(out_aod); free(out_h2o);
    return 0;
}

File: examples/03_retrievals.py
"""
03_retrievals.py — Atmospheric state retrievals.

Demonstrates:
  - Surface pressure from elevation (ISA)
  - H₂O column from 940 nm band depth
  - AOD from MODIS DDV + MAIAC patch-median smoothing
  - O₃ from Chappuis 600 nm band depth
  - Cloud/shadow/water/snow quality bitmask
  - Joint AOD + H₂O optimal estimation

Run:
    python3 03_retrievals.py
"""

import os, ctypes
import numpy as np

# ── Load libsixsv.so ──────────────────────────────────────────────────────────
_here = os.path.dirname(os.path.abspath(__file__))
for _p in [os.environ.get("LIB_SIXSV", ""),
           os.path.join(_here, "..", "testsuite", "libsixsv.so")]:
    if _p and os.path.exists(_p):
        _lib = ctypes.CDLL(_p); break
else:
    raise ImportError("libsixsv.so not found — run 'make' in testsuite/ first.")

_FP = ctypes.POINTER(ctypes.c_float)

class _Cfg(ctypes.Structure):
    _fields_ = [
        ("wl",  _FP), ("n_wl",  ctypes.c_int),
        ("aod", _FP), ("n_aod", ctypes.c_int),
        ("h2o", _FP), ("n_h2o", ctypes.c_int),
        ("sza", ctypes.c_float), ("vza", ctypes.c_float),
        ("raa", ctypes.c_float), ("altitude_km", ctypes.c_float),
        ("atmo_model",    ctypes.c_int), ("aerosol_model",    ctypes.c_int),
        ("surface_pressure", ctypes.c_float), ("ozone_du", ctypes.c_float),
        ("mie_r_mode",   ctypes.c_float), ("mie_sigma_g", ctypes.c_float),
        ("mie_m_real",   ctypes.c_float), ("mie_m_imag",  ctypes.c_float),
        ("brdf_type",    ctypes.c_int),   ("brdf_params",  ctypes.c_float * 5),
        ("enable_polar", ctypes.c_int),
    ]

class _Arr(ctypes.Structure):
    _fields_ = [
        ("R_atm", _FP), ("T_down", _FP), ("T_up", _FP), ("s_alb", _FP),
        ("T_down_dir", _FP), ("R_atmQ", _FP), ("R_atmU", _FP),
    ]

def _fp(a):
    return np.ascontiguousarray(a, dtype=np.float32).ctypes.data_as(_FP)

def compute_lut(wl, aod, h2o, sza=30., vza=5., raa=90., altitude_km=1000.,
                atmo_model=1, aerosol_model=1, ozone_du=300.,
                surface_pressure=0., enable_polar=0):
    wl_  = np.asarray(wl,  dtype=np.float32)
    aod_ = np.asarray(aod, dtype=np.float32)
    h2o_ = np.asarray(h2o, dtype=np.float32)
    N    = wl_.size * aod_.size * h2o_.size
    Ra, Td, Tu, ss = [np.empty(N, np.float32) for _ in range(4)]
    cfg_c = _Cfg(
        wl=_fp(wl_), n_wl=wl_.size, aod=_fp(aod_), n_aod=aod_.size,
        h2o=_fp(h2o_), n_h2o=h2o_.size,
        sza=sza, vza=vza, raa=raa, altitude_km=altitude_km,
        atmo_model=atmo_model, aerosol_model=aerosol_model,
        ozone_du=ozone_du, surface_pressure=surface_pressure,
        enable_polar=enable_polar,
    )
    arr_c = _Arr(R_atm=_fp(Ra), T_down=_fp(Td), T_up=_fp(Tu), s_alb=_fp(ss))
    _lib.atcorr_compute_lut(ctypes.byref(cfg_c), ctypes.byref(arr_c))
    sh = (aod_.size, h2o_.size, wl_.size)
    return dict(cfg_c=cfg_c, arr_c=arr_c,
                wl=wl_, aod=aod_, h2o=h2o_,
                R_atm=Ra.reshape(sh), T_down=Td.reshape(sh),
                T_up=Tu.reshape(sh),  s_alb=ss.reshape(sh),
                _Ra=Ra, _Td=Td, _Tu=Tu, _ss=ss)

# ── Quality-mask bit constants (from atcorr.h) ────────────────────────────────
MASK_CLOUD  = 0x01
MASK_SHADOW = 0x02
MASK_WATER  = 0x04
MASK_SNOW   = 0x08

# ── Scene setup ───────────────────────────────────────────────────────────────
npix = 32 * 32
doy, sza, vza = 185, 32.0, 5.0

# ── Synthetic per-pixel radiances ─────────────────────────────────────────────
def band(base, noise=5.0):
    return (base + noise * np.sin(np.arange(npix) * 0.37)).astype(np.float32)

L_470  = band(80.0, 5.0)
L_540  = band(70.0, 3.0)
L_600  = band(62.0, 3.0)
L_660  = band(55.0, 4.0)
L_680  = band(50.0, 3.0)
L_860  = band(120.0, 8.0)
L_865  = band(118.0, 8.0)
L_940  = (L_865 * 0.85 - 2.0 * np.sin(np.arange(npix) * 0.5)).astype(np.float32)
L_1040 = (L_865 * 0.92).astype(np.float32)
L_2130 = band(25.0, 2.0)

# ── 1. Surface pressure from elevation ────────────────────────────────────────
_lib.retrieve_pressure_isa.argtypes = [ctypes.c_float]
_lib.retrieve_pressure_isa.restype  = ctypes.c_float
p_surface = _lib.retrieve_pressure_isa(ctypes.c_float(1500.0))
print(f"Surface pressure at 1500 m: {p_surface:.1f} hPa")

# ── 2. H₂O retrieval from 940 nm band depth ───────────────────────────────────
_lib.retrieve_h2o_940.argtypes = [
    _FP, _FP, _FP,          # L_865, L_940, L_1040
    ctypes.c_float,          # fwhm_um (0 = broadband)
    ctypes.c_int,            # npix
    ctypes.c_float,          # sza_deg
    ctypes.c_float,          # vza_deg
    _FP,                     # out_wvc [npix]
]
_lib.retrieve_h2o_940.restype = None
wvc = np.empty(npix, np.float32)
_lib.retrieve_h2o_940(_fp(L_865), _fp(L_940), _fp(L_1040),
                      ctypes.c_float(0.), ctypes.c_int(npix),
                      ctypes.c_float(sza), ctypes.c_float(vza), _fp(wvc))
print(f"\nH₂O retrieval (940 nm): mean WVC = {wvc.mean():.2f} g/cm²")

# ── 3. AOD retrieval (DDV) ────────────────────────────────────────────────────
_lib.retrieve_aod_ddv.argtypes = [
    _FP, _FP, _FP, _FP,   # L_470, L_660, L_860, L_2130
    ctypes.c_int,           # npix
    ctypes.c_int,           # doy
    ctypes.c_float,         # sza_deg
    _FP,                    # out_aod [npix]
]
_lib.retrieve_aod_ddv.restype = ctypes.c_float
aod_px = np.empty(npix, np.float32)
aod_scene = _lib.retrieve_aod_ddv(
    _fp(L_470), _fp(L_660), _fp(L_860), _fp(L_2130),
    ctypes.c_int(npix), ctypes.c_int(doy), ctypes.c_float(sza), _fp(aod_px))
print(f"AOD retrieval (DDV): scene mean = {float(aod_scene):.3f}")

# MAIAC patch-median spatial regularisation (in-place, 32×32 grid)
_lib.retrieve_aod_maiac.argtypes = [
    _FP, ctypes.c_int, ctypes.c_int, ctypes.c_int]
_lib.retrieve_aod_maiac.restype = None
aod_2d = aod_px.reshape(32, 32).copy()
_lib.retrieve_aod_maiac(_fp(aod_2d), ctypes.c_int(32), ctypes.c_int(32),
                         ctypes.c_int(8))
print(f"AOD after MAIAC smoothing: mean = {aod_2d.mean():.3f}")

# ── 4. O₃ retrieval from Chappuis band ────────────────────────────────────────
_lib.retrieve_o3_chappuis.argtypes = [
    _FP, _FP, _FP,           # L_540, L_600, L_680
    ctypes.c_int,             # npix
    ctypes.c_float,           # sza_deg
    ctypes.c_float,           # vza_deg
]
_lib.retrieve_o3_chappuis.restype = ctypes.c_float
o3_du = _lib.retrieve_o3_chappuis(
    _fp(L_540), _fp(L_600), _fp(L_680),
    ctypes.c_int(npix), ctypes.c_float(sza), ctypes.c_float(vza))
print(f"O₃ column (Chappuis 600 nm): {float(o3_du):.0f} DU")

# ── 5. Quality bitmask ────────────────────────────────────────────────────────
_lib.retrieve_quality_mask.argtypes = [
    _FP, _FP, _FP,           # L_blue, L_red, L_nir
    _FP,                      # L_swir (or NULL for no snow)
    ctypes.c_int,             # npix
    ctypes.c_int,             # doy
    ctypes.c_float,           # sza_deg
    ctypes.POINTER(ctypes.c_uint8),  # out_mask [npix]
]
_lib.retrieve_quality_mask.restype = None
mask = np.zeros(npix, dtype=np.uint8)
_lib.retrieve_quality_mask(
    _fp(L_470), _fp(L_660), _fp(L_865),
    None,
    ctypes.c_int(npix), ctypes.c_int(doy), ctypes.c_float(sza),
    mask.ctypes.data_as(ctypes.POINTER(ctypes.c_uint8)))
n_cloud = int((mask & MASK_CLOUD).astype(bool).sum())
n_water = int((mask & MASK_WATER).astype(bool).sum())
print(f"Quality mask: cloud={n_cloud}  water={n_water}  (of {npix} pixels)")

# ── 6. Joint AOD + H₂O optimal estimation ─────────────────────────────────────
lut = compute_lut(
    wl  = [0.47, 0.55, 0.66, 0.87],
    aod = [0.0, 0.1, 0.2, 0.4],
    h2o = [1.0, 2.0, 3.0],
    sza=sza, vza=vza, raa=90., altitude_km=1000.,
    atmo_model=1, aerosol_model=1, ozone_du=300.,
)

n_vis   = 4
vis_wl  = np.array([0.47, 0.55, 0.66, 0.87], dtype=np.float32)
rho_vis = np.column_stack([
    (0.10 + 0.05 * b + 0.02 * np.sin(np.arange(npix) * 0.2)).astype(np.float32)
    for b in range(n_vis)
])  # [npix, n_vis]

_lib.oe_invert_aod_h2o.argtypes = [
    ctypes.POINTER(_Cfg), ctypes.POINTER(_Arr),
    _FP, ctypes.c_int, ctypes.c_int,   # rho_toa_vis, npix, n_vis
    _FP,                               # vis_wl
    _FP, _FP, _FP,                    # L_865, L_940, L_1040
    ctypes.c_float, ctypes.c_float,   # sza, vza
    ctypes.c_float, ctypes.c_float,   # aod_prior, h2o_prior
    ctypes.c_float, ctypes.c_float, ctypes.c_float,  # sigma_aod, _h2o, _spec
    ctypes.c_float,                   # fwhm_940
    _FP, _FP,                         # out_aod, out_h2o
]
_lib.oe_invert_aod_h2o.restype = None

out_aod = np.empty(npix, np.float32)
out_h2o = np.empty(npix, np.float32)
_lib.oe_invert_aod_h2o(
    ctypes.byref(lut['cfg_c']), ctypes.byref(lut['arr_c']),
    _fp(rho_vis), ctypes.c_int(npix), ctypes.c_int(n_vis), _fp(vis_wl),
    _fp(L_865), _fp(L_940), _fp(L_1040),
    ctypes.c_float(sza), ctypes.c_float(vza),
    ctypes.c_float(float(aod_scene)), ctypes.c_float(float(wvc.mean())),
    ctypes.c_float(0.5), ctypes.c_float(1.0), ctypes.c_float(0.01),
    ctypes.c_float(0.),
    _fp(out_aod), _fp(out_h2o))
print(f"OE retrieval: mean AOD = {out_aod.mean():.3f}  "
      f"mean H₂O = {out_h2o.mean():.2f} g/cm²")
